File: Source/ChilliSource/Core/Math/Quaternion.cpp

#include <moFlo/Core/Math/Quaternion.h>
#include <moFlo/Core/Math/MathUtils.h>

using namespace moFlo::Core;

const CQuaternion CQuaternion::IDENTITY(0,0,0,1);
const CQuaternion CQuaternion::ZERO(0,0,0,0);

CQuaternion::CQuaternion()
:x(0),y(0),z(0),w(1)
{
	
}

CQuaternion::CQuaternion(const f32 _x, const f32 _y, const f32 _z, const f32 _w)
:w(_w),x(_x),y(_y),z(_z)
{
}

CQuaternion::CQuaternion(const CVector3 & V, f32 theta)
{
	SetAxisAngle(V,theta);
}
//-----------------------------------------------------------------------
CQuaternion::CQuaternion(const CVector3& xaxis, const CVector3& yaxis, const CVector3& zaxis)
{
	CMatrix4x4 matRot;
	
	matRot.m[0] = xaxis.x;
	matRot.m[1] = xaxis.y;
	matRot.m[2] = xaxis.z;
	
	matRot.m[4] = yaxis.x;
	matRot.m[5] = yaxis.y;
	matRot.m[6] = yaxis.z;
	
	matRot.m[8] = zaxis.x;
	matRot.m[9] = zaxis.y;
	matRot.m[10] = zaxis.z;
	
	(*this) = CQuaternion(matRot);
}
f32 CQuaternion::GetAngle() const {
	f32 fSqrLength = (x * x + y * y + z * z);

	if (fSqrLength > 0){
		return 2.0f * std::acos(w);
	} 

	return 0.0f;
}
CVector3 CQuaternion::GetAxis() const {
	f32 fSqrLength = (x * x + y * y + z * z);

	if (fSqrLength > 0){
		f32 fInvLength = 1.0f / std::sqrt(fSqrLength);
		return CVector3(x * fInvLength, y * fInvLength, z * fInvLength);
	} else {
		return CVector3::X_UNIT_POSITIVE;
	}
}

void CQuaternion::ToEulerAxes (CVector3& xaxis, CVector3& yaxis, CVector3& zaxis) const
{
	CMatrix4x4 matRot;
	
	ToRotationMatrix(matRot);
	
	xaxis.x = matRot(0, 0);
	xaxis.y = matRot(0, 1);
	xaxis.z = matRot(0, 2);
	
	yaxis.x = matRot(1, 0);
	yaxis.y = matRot(1, 1);
	yaxis.z = matRot(1, 2);
	
	zaxis.x = matRot(2, 0);
	zaxis.y = matRot(2, 1);
	zaxis.z = matRot(2, 2);
}
CQuaternion::CQuaternion(const CMatrix4x4& inMat)
{
	// Algorithm in Ken Shoemake's article in 1987 SIGGRAPH course notes
	// article "Quaternion Calculus and Fast Animation".
	
	f32 fTrace = inMat.m[0]+inMat.m[5]+inMat.m[10];
	f32 fRoot = 0.0f;
	
	if (fTrace > 0.0)
	{
		// |w| > 1/2, may as well choose w > 1/2
		fRoot = sqrtf(fTrace + 1.0f);  // 2w
		w = 0.5f*fRoot;
		fRoot = 0.5f/fRoot;  // 1/(4w)
		x = (inMat.m[6]-inMat.m[9])*fRoot;
		y = (inMat.m[8]-inMat.m[2])*fRoot;
		z = (inMat.m[1]-inMat.m[4])*fRoot;
	}
	else
	{
		// |w| <= 1/2
		static size_t s_iNext[3] = { 1, 2, 0 };
		size_t i = 0;
		if ( inMat.m[5] > inMat.m[0] )
			i = 1;
		if ( inMat.m[10] > inMat(i,i) )
			i = 2;
		size_t j = s_iNext[i];
		size_t k = s_iNext[j];
		
		fRoot = sqrtf(inMat(i,i)-inMat(j,j)-inMat(k,k) + 1.0f);
		
		f32* apkQuat[3] = { &x, &y, &z };
		*apkQuat[i] = 0.5f*fRoot;
		fRoot = 0.5f/fRoot;
		w = (inMat(j,k)-inMat(k,j))*fRoot;
		*apkQuat[j] = (inMat(j,i)+inMat(i,j))*fRoot;
		*apkQuat[k] = (inMat(k,i)+inMat(i,k))*fRoot;
	}
}
void CQuaternion::SetAxisAngle(const CVector3 &V, f32 theta)
{
	CVector3 Vnorm = V.NormalisedCopy();

	f32 fHalfAngle = theta/2.0f;
	f32 fSinAngle = (f32)sin(fHalfAngle);
	x = Vnorm.x * fSinAngle;
	y = Vnorm.y * fSinAngle;
	z = Vnorm.z * fSinAngle;
	w = (f32)cos(fHalfAngle);
}
void CQuaternion::SetAxisAngle(f32 infX, f32 infY, f32 infZ, f32 theta)
{
	f32 fHalfAngle = theta/2.0f;
	f32 fSinAngle = (f32)sin(fHalfAngle);
	x = infX * fSinAngle;
	y = infY * fSinAngle;
	z = infZ * fSinAngle;
	w = (f32)cos(fHalfAngle);
}

CQuaternion CQuaternion::Slerp(const CQuaternion &q1, const CQuaternion &q2, float t, bool inbShortestPath) 
{
	CQuaternion result, _q2 = q2;
	
	float cosOmega = q1.w*q2.w + q1.x*q2.x + q1.y*q2.y + q1.z*q2.z;
	
	if ( cosOmega < 0.0f && inbShortestPath ) 
	{
		_q2.x = -_q2.x;
		_q2.y = -_q2.y;
		_q2.z = -_q2.z;
		_q2.w = -_q2.w;
		cosOmega = -cosOmega;
	}
	
	float k0, k1;
	if ( cosOmega > 0.99999f ) {
		k0 = 1.0f - t;
		k1 = t;
		
		result.x = q1.x*k0 + _q2.x*k1;
		result.y = q1.y*k0 + _q2.y*k1;
		result.z = q1.z*k0 + _q2.z*k1;
		result.w = q1.w*k0 + _q2.w*k1;
		
		result.NormaliseSelf();		
	}
	else {
		float sinOmega = (float)sqrt( double(1.0f - cosOmega*cosOmega) );
		float omega = (float)atan2( double(sinOmega), double(cosOmega) );
		
		float invSinOmega = 1.0f/sinOmega;
		
		k0 = float( sin(double((1.0f - t)*omega)) )*invSinOmega;
		k1 = float( sin(double(t*omega)) )*invSinOmega;
		
		result.x = q1.x*k0 + _q2.x*k1;
		result.y = q1.y*k0 + _q2.y*k1;
		result.z = q1.z*k0 + _q2.z*k1;
		result.w = q1.w*k0 + _q2.w*k1;
	}
	return result;
}

CQuaternion CQuaternion::NLerp(const moFlo::Core::CQuaternion &q1, const moFlo::Core::CQuaternion &q2, float t, bool inbShortestPath)
{
    const f32 fDot = q1.x * q2.x + q1.y * q2.y + q1.z * q2.z + q1.x * q2.x + q1.w * q2.w;
    CQuaternion rVal;
    if(fDot < 0.0f && inbShortestPath)
        rVal = q1 + (t * ((q2 * -1.0f) - q1));
    else
        rVal = q1 + (t * (q2 -q1));
    
    rVal.NormaliseSelf();
    return rVal;
}

CQuaternion::~CQuaternion()
{
}

f32 CQuaternion::Modulus() const
{
    return sqrtf(w * w + x * x + y * y + z * z);
}

f32 CQuaternion::ModulusSquared() const
{
    return (w * w + x * x + y * y + z * z);
}

CQuaternion & CQuaternion::NormaliseSelf()
{
	f32 s = 1.0f / Modulus();
	w *= s;
	x *= s;
	y *= s;
	z *= s;

	return *this;
}

CQuaternion CQuaternion::Conjugate() const
{
	CQuaternion quat;
	quat.x = -x;
	quat.y = -y;
	quat.z = -z;
	quat.w = w;
	return quat;
}

CQuaternion CQuaternion::Inverse() const
{
    return Conjugate() / ModulusSquared();
}

void CQuaternion::ToRotationMatrix(CMatrix4x4 & inMatrix) const
{
	f32 x2 = x * x;
	f32 y2 = y * y;
	f32 z2 = z * z;
	f32 xy = x * y;
	f32 xz = x * z;
	f32 yz = y * z;
	f32 wx = w * x;
	f32 wy = w * y;
	f32 wz = w * z;
	
	inMatrix = CMatrix4x4
	( 
	 1.0f - 2.0f * (y2 + z2),	2.0f * (xy + wz),			2.0f * (xz - wy),			0.0f,
	 2.0f * (xy - wz),			1.0f - 2.0f * (x2 + z2),	2.0f * (yz + wx),			0.0f,
	 2.0f * (xz + wy),			2.0f * (yz - wx),			1.0f - 2.0f * (x2 + y2),	0.0f,
	 0.0f,						0.0f,						0.0f,						1.0f
	 );
}

CMatrix4x4 CQuaternion::ToRotationMatrix4x4() const
{

	CMatrix4x4 Result;
	this->ToRotationMatrix(Result);

	return Result;
}

File: Projects/Tools/CoreUtils/src/com/chillisource/toolutils/Quaternion.java
/**
 * Quaternion.java
 * Chilli Source
 * Created by Ian Copland on 30/01/2013
 * 
 * The MIT License (MIT)
 * 
 * Copyright (c) 2013 Tag Games Limited
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.chillisource.toolutils;

public class Quaternion 
{
	//------------------------------------------------------------
	/// Public member data
	//------------------------------------------------------------
	public float x;
	public float y;
	public float z;
	public float w;
	//------------------------------------------------------------
	/// Constructor
	//------------------------------------------------------------
	public Quaternion()
	{
		x = y = z = w = 0.0f;
	}
	//------------------------------------------------------------
	/// Set
	///
	/// Sets this quaternion to the value of the given quaternion.
	///
	/// @param the quaternion to copy.
	//------------------------------------------------------------
	public void set(Quaternion inQuat)
	{
		x = inQuat.x;
		y = inQuat.y;
		z = inQuat.z;
		w = inQuat.w;
	}
	//------------------------------------------------------------
	/// To Matrix
	///
	/// @return the matrix equivalent of this quaternion.
	//------------------------------------------------------------
	public Matrix4 toMatrix()
	{
		float x2 = x * x;
		float y2 = y * y;
		float z2 = z * z;
		float xy = x * y;
		float xz = x * z;
		float yz = y * z;
		float wx = w * x;
		float wy = w * y;
		float wz = w * z;
		
		Matrix4 output = new Matrix4();
		output.mafData[0] = 1.0f - 2.0f * (y2 + z2);	output.mafData[1] = 2.0f * (xy + wz);			output.mafData[2] = 2.0f * (xz - wy);			output.mafData[3] = 0.0f;
		output.mafData[4] = 2.0f * (xy - wz);			output.mafData[5] = 1.0f - 2.0f * (x2 + z2);	output.mafData[6] = 2.0f * (yz + wx);			output.mafData[7] = 0.0f;
		output.mafData[8] = 2.0f * (xz + wy);			output.mafData[9] = 2.0f * (yz - wx);			output.mafData[10] = 1.0f - 2.0f * (x2 + y2);	output.mafData[11] = 0.0f;
		output.mafData[12] = 0.0f;						output.mafData[13] = 0.0f;						output.mafData[14] = 0.0f;						output.mafData[15] = 1.0f;
		
		return output;
	}
	//------------------------------------------------------------
	/// Get Axis
	///
	/// @return the axis
	//------------------------------------------------------------
	public Vector3 getAxis()
	{
		float fSqrLength = (x * x + y * y + z * z);

		if (fSqrLength > 0.0f)
		{
			float fInvLength = 1.0f / (float)Math.sqrt(fSqrLength);
			return new Vector3(x * fInvLength, y * fInvLength, z * fInvLength);
		}

		return new Vector3(1.0f, 0.0f, 0.0f);
	}
	//------------------------------------------------------------
	/// Get Angle
	///
	/// @return the angle.
	//------------------------------------------------------------
	public float getAngle()
	{
		float fSqrLength = (x * x + y * y + z * z);

		if (fSqrLength > 0.0f)
		{
			return 2.0f * (float)Math.acos(w);
		} 

		return 0.0f;
	}
	//------------------------------------------------------------
	/// Normalise
	///
	/// @return a normalised version of this quaternion
	//------------------------------------------------------------
	public Quaternion normalise()
	{
		Quaternion output = new Quaternion();
		float s = 1.0f / (float)Math.sqrt(w * w + x * x + y * y + z * z);
		output.w = w * s;
		output.x = x * s;
		output.y = y * s;
		output.z = z * s;

		return output;
	}
	//------------------------------------------------------------
	/// Create From Matrix
	///
	/// Converts a rotation matrix into a quaternion.
	///
	/// @param the rotation matrix.
	/// @return the quaterion.
	//------------------------------------------------------------
	public static Quaternion createFromMatrix(Matrix4 inMat)
	{
		// Algorithm in Ken Shoemake's article in 1987 SIGGRAPH course notes
		// article "Quaternion Calculus and Fast Animation".
		
		Quaternion newQuat = new Quaternion();
		
		float fTrace = inMat.mafData[0]+inMat.mafData[5]+inMat.mafData[10] + 1.0f;
		float fRoot = 0.0f;
		
		if (fTrace > 0.0)
		{
			// |w| > 1/2, may as well choose w > 1/2
			fRoot = (float)Math.sqrt(fTrace);  // 2w
			newQuat.w = 0.5f*fRoot;
			fRoot = 0.5f/fRoot;  // 1/(4w)
			newQuat.x = (inMat.mafData[6]-inMat.mafData[9])*fRoot;
			newQuat.y = (inMat.mafData[8]-inMat.mafData[2])*fRoot;
			newQuat.z = (inMat.mafData[1]-inMat.mafData[4])*fRoot;
		}
		else
		{
			// |w| <= 1/2
			int[] s_iNext = { 1, 2, 0 };
			int i = 0;
			if ( inMat.mafData[5] > inMat.mafData[0] )
				i = 1;
			if ( inMat.mafData[10] > inMat.get(i,i) )
				i = 2;
			int j = s_iNext[i];
			int k = s_iNext[j];
			
			fRoot = (float)Math.sqrt(inMat.get(i,i)-inMat.get(j,j)-inMat.get(k,k) + 1.0f);
			
			float[] apkQuat = {0.0f, 0.0f, 0.0f};
			apkQuat[i] = 0.5f*fRoot;
			fRoot = 0.5f/fRoot;
			newQuat.w = (inMat.get(k,j)-inMat.get(j,k))*fRoot;
			apkQuat[j] = (inMat.get(j,i)+inMat.get(i,j))*fRoot;
			apkQuat[k] = (inMat.get(k,i)+inMat.get(i,k))*fRoot;
			
			newQuat.x = apkQuat[0];
			newQuat.y = apkQuat[1];
			newQuat.z = apkQuat[2];
		}

		return newQuat;
	}
	//------------------------------------------------------------
	/// Create From Axis Angle
	///
	/// Creates a new quaternion from an axis and an angle.
	///
	/// @param the axis.
	/// @param the angle.
	/// @return the quaterion.
	//------------------------------------------------------------
	public static Quaternion createFromAxisAngle(Vector3 invAxis, float infAngle)
	{
		Quaternion output = new Quaternion();
		Vector3 Vnorm = invAxis.normalise();

		float fHalfAngle = infAngle / 2.0f;
		float fSinAngle = (float)Math.sin(fHalfAngle);
		output.x = Vnorm.x * fSinAngle;
		output.y = Vnorm.y * fSinAngle;
		output.z = Vnorm.z * fSinAngle;
		output.w = (float)Math.cos(fHalfAngle);
		
		return output;
	}
}
